lib/ecto/changeset/relation.ex

```diff
@@ -295,7 +310,10 @@ defmodule Ecto.Changeset.Relation do
     {current_pks, current_map} = process_current(current, current_pks_fun, relation)
     %{unique: unique, ordered: ordered} = relation
     ordered = if ordered, do: current_pks, else: []
-    map_changes(value, new_pks_fun, fun, current_map, [], true, true, unique && %{}, ordered)
+
+    value
+    |> Enum.with_index()
+    |> map_changes(new_pks_fun, fun, current_map, [], true, true, unique && %{}, ordered)
```

Comment:
I'm computing `Enum.with_index` for all associations, even the ones that aren't using this feature. Probably this could be avoided, but would imply deeper refactor.

Reply:
Given `map_changes` is recursive, you could pass `idx` there and increment it yourself on each iteration?

Reply:
great call, i missed it! fixed :)